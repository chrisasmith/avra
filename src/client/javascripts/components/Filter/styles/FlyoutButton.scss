@import '../../../../stylesheets/modules/theme.scss';

.flyoutButton {
  background-color: #60636f;
  color: #fff;
  height: 34px;
  line-height: 34px;
  font-size: 16px;
  cursor: pointer;

  &.open, &:hover {
    background-color: lighten(#60636f, 40%);
    color: #1d1f2a;
  }

  &.disabled {
    background-color: #c0c0c0;
    cursor: hand;

    pointer-events: none;
  }
}

.rightArrow {
  float: right;
  line-height: 34px;
  margin-right: 16px;
}
