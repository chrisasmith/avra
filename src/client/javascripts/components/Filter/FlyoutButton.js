import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import styles from './styles/FlyoutButton.scss'
import FlyoutMenu from './FlyoutMenu'

const propTypes = {
  options: PropTypes.array.isRequired,
  onItemSelected: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  className: PropTypes.string,
  disabled: PropTypes.bool,
}

class FlyoutButton extends Component {
  constructor(props, context) {
    super(props, context)

    this.onButtonClick = this.onButtonClick.bind(this)
    this.onItemSelected = this.onItemSelected.bind(this)

    this.state = {
      flyoutOpen: false,
    }
  }

  onItemSelected(item) {
    this.setState({ flyoutOpen: false })

    if (item) {
      // remove the children
      this.props.onItemSelected({ ...item, children: undefined })
    }
  }

  onButtonClick() {
    this.setState({ flyoutOpen: true })
  }

  render() {
    return (
      <div className={this.props.className}>
        {this.state.flyoutOpen && (
          <FlyoutMenu
            options={this.props.options}
            onItemSelected={this.onItemSelected}
          />
        )}
        <div
          className={classnames('pl-3', styles.flyoutButton, {
            [styles.open]: this.state.flyoutOpen,
            [styles.disabled]: this.props.disabled,
          })}
          onClick={this.onButtonClick}
        >
          <i className={classnames('fa fa-chevron-right', styles.rightArrow)} />
          {this.props.label}
        </div>
      </div>
    )
  }
}

FlyoutButton.propTypes = propTypes
export default FlyoutButton
