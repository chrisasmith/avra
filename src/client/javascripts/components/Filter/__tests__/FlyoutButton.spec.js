import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import FlyoutButton from '../FlyoutButton'

describe('Components FlyoutButton', () => {
  it('should render', () => {
    const wrapper = shallow(<FlyoutButton
      options={[]}
      onItemSelected={jest.fn()}
      label="label"
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should render disabled', () => {
    const wrapper = shallow(<FlyoutButton
      options={[]}
      onItemSelected={jest.fn()}
      label="label"
      disabled
    />)

    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('should open the flyout', () => {
    const wrapper = shallow(<FlyoutButton
      options={[]}
      onItemSelected={jest.fn()}
      label="label"
    />)

    wrapper.find('.flyoutButton').simulate('click')
    expect(wrapper.state('flyoutOpen')).toBeTruthy()
  })

  it('should close the flyout', () => {
    const wrapper = shallow(<FlyoutButton
      options={[]}
      onItemSelected={jest.fn()}
      label="label"
    />)

    wrapper.find('.flyoutButton').simulate('click')
    expect(wrapper.state('flyoutOpen')).toBeTruthy()

    wrapper.instance().onItemSelected()
    expect(wrapper.state('flyoutOpen')).toBeFalsy()
  })

  it('should select an item', () => {
    const onItemSelected = jest.fn()
    const item = { name: 'hi', id: 1 }
    const wrapper = shallow(<FlyoutButton
      options={[item]}
      onItemSelected={onItemSelected}
      label="label"
    />)

    wrapper.find('.flyoutButton').simulate('click')
    expect(wrapper.state('flyoutOpen')).toBeTruthy()

    wrapper.instance().onItemSelected(item)
    expect(onItemSelected).toHaveBeenCalledWith(item)
  })
})
