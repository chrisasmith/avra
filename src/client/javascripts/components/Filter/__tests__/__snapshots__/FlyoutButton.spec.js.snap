// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components FlyoutButton should render 1`] = `
<div>
  <div
    className="pl-3 flyoutButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-chevron-right rightArrow"
    />
    label
  </div>
</div>
`;

exports[`Components FlyoutButton should render disabled 1`] = `
<div>
  <div
    className="pl-3 flyoutButton disabled"
    onClick={[Function]}
  >
    <i
      className="fa fa-chevron-right rightArrow"
    />
    label
  </div>
</div>
`;
